src/main/java/com/google/devtools/build/lib/query2/PostAnalysisQueryEnvironment.java

```diff
@@ -575,6 +579,51 @@ public Collection<T> getSiblingTargetsInPackage(T target) throws QueryException
   @Override
   public void close() {}
 
+  protected abstract OrderOutput getOrderOutput();
+
+  protected abstract Comparator<T> getFullOrderingComparator();
+
+  public Iterable<T> orderResults(Iterable<T> resultList)
+          throws InterruptedException {
+    OrderOutput orderOutput = this.getOrderOutput();
+
+    // No-op if order_output=no
+    if (orderOutput == OrderOutput.NO) {
+      return resultList;
+    }
+
+    // Behavior of order_output=auto depends on the output format
+    if (orderOutput == OrderOutput.AUTO) {
+      orderOutput = this.getOutputFormat().endsWith("proto") ? OrderOutput.DEPS : OrderOutput.FULL;
+    }
+
+    Digraph<T> graph = new Digraph<>();
+
+    // First pass: create nodes for dependency Digraph
+    for (T result : resultList) {
+      graph.createNode(result);
+    }
+
+    // Second pass: create edges between nodes
+    for (T result : resultList) {
+      for (T dep : getFwdDeps(ImmutableList.of(result))) {
+        if (graph.getNodeMaybe(dep) != null) {
+          graph.addEdge(
+                  graph.getNode(result),
+                  graph.getNode(dep)
```

Comment:
Do the results actually match the original list?

Ultimately we want the same nodes here as were in `resultList`. don't we? What happens if `resultList` is only a subset of the original graph? Like it applies a `deps` function but also a `kind` function to filter to only certain rule kinds?

If that is a thing, one option could be to post-filter the graph output to only those nodes in `resultList`.

Reply:
> Do the results actually match the original list?

I believe so. The only place nodes are created is in the block above (lines 603-605) where one node is created per item in `resultList`. Here we're just adding edges between nodes but not creating any new ones.

Reply:
Oh, I see. Line 610 filters out the possibility of extra nodes?

What happens if the query applies filtering down the transitive closure? Like `kind("cc_library", deps(//foo))`? Where `//foo` has no direct deps on `cc_library` targets but has transitive ones?

Reply:
In that case the ordering between `//foo` and the `cc_library` targets would not be preserved because the intermediate targets would not be present in `resultList`. If we need to preserve the pre-filtering ordering we will have to move this logic to an earlier point in the query (which would also address [this issue](https://github.com/bazelbuild/bazel/pull/16427/files/2b8a1c1e9c87b5b263a4264bc95d472a4b3056f3#r1003593251)).

Reply:
My concern is that we want to be able to consolidate what's already complex and subtle behavior. The main documentation - at https://bazel.build/query/language#results-ordering - we shouldn't skew too much in the cquery / aquery case. If cquery/aquery only supports a subset of the settings that's fine. But if each value has a different interpretation that's going to get confusing.

It's a naive followup, but can you clarify your original use case? That you needed a basic topological ordering to be able to do diffs, to identify changes and which tests are relevant? What does a simple example look like?

Reply:
Yeah that makes sense. Test selection was the initial motivation for adding this support but that's actually no longer an issue for us. As this PR would need some major changes that I don't have bandwidth to work on now I'll close it out.